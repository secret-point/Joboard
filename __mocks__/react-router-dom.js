// Mock navigate and useParams when needed
// const navigate = jest.fn();
// const useParams = jest.fn();
const useLocation = jest.fn();

module.exports = {
  ...jest.requireActual("react-router-dom"),
  // useNavigate: () => navigate,
  // useParams,
  useLocation,
};
