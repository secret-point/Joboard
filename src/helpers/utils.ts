import isEmpty from "lodash/isEmpty";

export const convertPramsToJson = (params: string) => {
  if (!isEmpty(params)) {
    return JSON.parse(
      '{"' +
        decodeURI(params.substring(1))
          .replace(/"/g, '\\"')
          .replace(/&/g, '","')
          .replace(/=/g, '":"') +
        '"}'
    );
  } else {
    return {};
  }
};

export const launchAuthentication = () => {
  let hash = window.location.hash.substr(1);

  const origin = window.location.origin;
  const redirectUrl = `${origin}/#${hash}`;

  const state = window.reduxStore.getState();
  let url = `${
    state.app.appConfig.authenticationURL
  }/?redirectUrl=${encodeURIComponent(redirectUrl)}`;
  window.location.assign(url);
};

export const isJson = (obj: string) => {
  try {
    JSON.parse(obj);
  } catch (e) {
    return false;
  }
  return true;
};
