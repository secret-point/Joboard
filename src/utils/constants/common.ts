export const HVH_LOCALE = 'hvh-locale';
