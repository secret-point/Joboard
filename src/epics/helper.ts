import { log, LoggerType } from "../helpers/log-helper";
import { ProxyApiError } from "../utils/api/types";

export const epicSwitchMapHelper = async (response: any) => {
  if (!response.data || !!response.errorCode || !!response.error || response.errorMessage) {
    log("[Epic] epicSwitchMapHelper: Starting to raise proxy api error", response, LoggerType.ERROR);
    const { errorCode } = response;
    throw createProxyApiEpicError(errorCode, response.error || response.errorMessage);
  }
  return response;
};

export const createProxyApiEpicError = (errorCode: string, errorMessage?: string): ProxyApiError => {
  return { errorCode, errorMessage };
};
