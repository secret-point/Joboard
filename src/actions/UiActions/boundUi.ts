import store from "../../store/store";
import { actionGetCountryStateConfig, actionResetBannerMessage, actionSetBannerMessage } from "./uiActions";
import { AlertMessage } from "../../utils/types/common";

export const boundGetCountryStateConfig = () => store.dispatch(actionGetCountryStateConfig());
export const boundSetBannerMessage = (payload: AlertMessage) => store.dispatch(actionSetBannerMessage(payload));
export const boundResetBannerMessage = () => store.dispatch(actionResetBannerMessage());

