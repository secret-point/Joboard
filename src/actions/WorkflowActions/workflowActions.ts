import moment from "moment";
import {
  PAGE_ROUTES,
  PagesControlledByWorkFlowService,
  PagesNeedToUseWorkflowRedirection
} from "../../components/pageRoutes";
import { MAX_MINUTES_FOR_HEARTBEAT } from "../../constants";
import { log, logError } from "../../helpers/log-helper";
import { checkIfIsCSRequest } from "../../helpers/utils";
import StepFunctionService from "../../services/step-function-service";
import store from "../../store/store";
import { WORKFLOW_ERROR_CODE, WORKFLOW_STEP_NAME } from "../../utils/enums/common";
import { getCurrentStepNameFromHash, loadingStatusHelper, routeToAppPageWithPath } from "../../utils/helper";
import { Application, CompleteTaskRequest, EnvConfig, Schedule, WorkflowData } from "../../utils/types/common";
import { boundUpdateWorkflowName } from "../ApplicationActions/boundApplicationActions";
import { boundWorkflowRequestEnd, boundWorkflowRequestStart } from "../WorkflowActions/boundWorkflowActions";
import {
  SetWorkflowErrorCodeAction,
  WORKFLOW_REQUEST,
  WorkflowRequestEndAction,
  WorkflowRequestInitAction,
  WorkflowRequestStartAction
} from "./workflowActionTypes";

export const loadWorkflow =
    ( requisitionId: string, applicationId: string, candidateId: string, envConfig: EnvConfig, isCompleteTaskOnLoad?: boolean, applicationData?: Application ) => {
      if (!window?.stepFunctionService?.websocket) {

        if (isCompleteTaskOnLoad) {
          window.isCompleteTaskOnLoad = isCompleteTaskOnLoad;
          window.applicationData = applicationData;
        }

        log("Initiated to connect websocket");
        window.stepFunctionService = StepFunctionService.load(requisitionId, applicationId, candidateId, envConfig);
      }
    };

export const loadWorkflowDS =
    ( jobId: string, scheduleId: string, applicationId: string, candidateId: string, envConfig: EnvConfig, isCompleteTaskOnLoad?: boolean, applicationData?: Application ) => {
      if (!window?.stepFunctionService?.websocket) {

        if (isCompleteTaskOnLoad) {
          window.isCompleteTaskOnLoad = isCompleteTaskOnLoad;
          window.applicationData = applicationData;
        }

        log("[WS] Initiated to connect websocket", [jobId, scheduleId, applicationId, candidateId, envConfig]);
        window.stepFunctionService = StepFunctionService.loadDS(jobId, scheduleId, applicationId, candidateId, envConfig);
      }
    };

export const startOrResumeWorkflow = () => {
  log("Started workflow");
  const { stepFunctionService } = window;
  const { applicationId, candidateId, requisitionId } = stepFunctionService;
  const payload: string = JSON.stringify({
    action: "startWorkflow",
    applicationId,
    candidateId,
    requisitionId,
    isCsDomain: checkIfIsCSRequest()
  });
  window.stepFunctionService.sendMessage(payload);
};

export const startOrResumeWorkflowDS = () => {
  log("[WS] Started DS workflow");
  boundWorkflowRequestStart();
  const { stepFunctionService } = window;
  const { applicationId, candidateId, jobId, scheduleId } = stepFunctionService;

  const payload: string = JSON.stringify({
    action: "startWorkflow",
    applicationId,
    candidateId,
    jobId,
    scheduleId,
    isCsDomain: checkIfIsCSRequest()
  });

  window.stepFunctionService.sendMessage(payload);

  if (window.hasCompleteTask) {
    window.hasCompleteTask();
    window.hasCompleteTask = undefined;
  }
};

export const sendHeartBeatWorkflow = () => {
  const { websocket } = window.stepFunctionService;
  if (window.hearBeatTime) {
    log("Sending the heart beat event");
    const endTime = moment();
    const startTime = moment(window.hearBeatTime);
    const duration = moment.duration(endTime.diff(startTime));

    if (duration.asMinutes() < MAX_MINUTES_FOR_HEARTBEAT && websocket?.OPEN === websocket?.readyState) {
      const payload: string = JSON.stringify({ action: "heartbeat" });
      window.stepFunctionService.sendMessage(payload);
    } else {
      log("Websocket timed out, moved to timed out page");
      routeToAppPageWithPath(PAGE_ROUTES.TIMEOUT);
    }
  } else {
    window.hearBeatTime = moment().toISOString();

    if (websocket?.OPEN === websocket?.readyState) {
      log("Sending the heart beat event");
      const payload: string = JSON.stringify({ action: "heartbeat" });
      window.stepFunctionService.sendMessage(payload);
    } else {
      log("Websocket timed out, moved to timed out page");
      routeToAppPageWithPath(PAGE_ROUTES.TIMEOUT);
    }
  }
};

export const ifShouldGoToStep = (targetStepName: string, currentStepName: string): boolean => {

  if (targetStepName === currentStepName) {
    // Do not redirect if the current step is the same as the target step

    return false;
  } else if (currentStepName === PAGE_ROUTES.ASSESSMENT_FINISHED &&
    targetStepName === WORKFLOW_STEP_NAME.ASSESSMENT_CONSENT) {

    // Do not redirect to `assessment-consent` if the current step is `assessment-finished`.
    //
    // There is a race condition on `assessment-finished` page, when HOOK redirect back to us, since the whole domain and url change it's a
    // page refresh. BB UI will reconnect websocket which will returns the step `assessment-consent`, since there is a mismatch between the
    // current step and the current page, BB UI will try to redirect the user to the `assessment-consent` page the WS response indicates.
    //
    // We want to make sure that the user stays at `assessment-finished` page when we are trying to completeTask assessment-consent. Otherwise,
    // the user will see the `assessment-consent` page first then after a few seconds be redirected back to the next step.

    return false;
  } else if (currentStepName === PAGE_ROUTES.WOTC_COMPLETE &&
    targetStepName === WORKFLOW_STEP_NAME.WOTC) {

    // Make sure that the user stays at `wotc-complete` page when we are trying to completeTask wotc. Otherwise,
    // the user will see the `wotc` page first then after a few seconds be redirected back to the next step, which is `supplementary-success`.
    // Right now wotc page takes a long time (~2-5s) to load that's why we didn't see any page flickering at this time.

    return false;
  } else if (Object.values(PAGE_ROUTES).includes(currentStepName as PAGE_ROUTES) &&
    !Object.values(PagesControlledByWorkFlowService).includes(currentStepName as PAGE_ROUTES)
    && !Object.values(PagesNeedToUseWorkflowRedirection).includes(currentStepName as PAGE_ROUTES) && !window.hasTargetStep) {
    // this is to address the infinite redirection for any error page that fetch application from proxy. https://sim.amazon.com/issues/Kondo_QA_Issue-121
    // Typically if an error page get application detail and trigger GetApplicationSuccessEpic if there is no websocket connection setup,
    // it will start workflow connection, once workflow connection is started it will return the nextStep which is different to error page
    // as error pages are not managed by workflow service. In this scenario, we redirect to the page returned by workflow service which shouldn't be the right one
    // With this change, we will check if a page is a valid route and is not managed by workflow service, then we don't have to redirect to the targeted page from workflow service
    // instead we stay on this error page to avoid probable infinite loop
    // If a page is invalid route, or is managed by workflow service, we rely on it
    // we should exclude consent page to ensure that redirection to job opportunity or assessment works.
    // check if it has targetStep to ensure that when we are on error page and try to go to workflow service managed page, this case we can rely on workflow service nextStep from response

    return false;
  }
  // reset hasTarget to ensure next page will be redirected correctly by relying on workflow service nextStep if it's on a UI controlled page
  window.hasTargetStep = undefined;
  return true;
  
};

export const goToStep = async ( workflowData: WorkflowData ) => {
  const state = store.getState();
  const applicationData = state.application.results;
  const currentStepName = getCurrentStepNameFromHash();
  const { stepName } = workflowData;

  log("Received data from step function", { workflowData, currentStepName: currentStepName, goToStepName: stepName });

  if (stepName && applicationData && ifShouldGoToStep(stepName, currentStepName)) {
    boundWorkflowRequestStart();

    log(`current step name (${currentStepName}) and go to step name (${stepName}) is not matched`);
    log("updating workflow step name in application", { applicationId: applicationData.applicationId, stepName });

    boundUpdateWorkflowName({ applicationId: applicationData.applicationId, workflowStepName: stepName }, () => {
      boundWorkflowRequestEnd();
      routeToAppPageWithPath(stepName);
      log(`update workflow step in local storage as ${stepName}`, applicationData);
      window.localStorage.setItem("page", stepName);
    }, ( ex: any ) => {
      logError("Unable to update workflow step in application", ex, applicationData);
      boundWorkflowRequestEnd();
    }
    );

    if (workflowData.stepName === WORKFLOW_STEP_NAME.SUPPLEMENTARY_SUCCESS || workflowData.stepName === WORKFLOW_STEP_NAME.THANK_YOU) {
      const metric = window.MetricsPublisher.newChildActionPublisherForMethod("ApplicationCompleteTime");
      const metricName = workflowData.stepName === WORKFLOW_STEP_NAME.SUPPLEMENTARY_SUCCESS ? "Completed" : "PreHireStepsCompleted";
      metric.publishTimerMonitor(metricName, Date.now() - window.applicationStartTime);
    }
  } else {
    log(`Received target step name: ${stepName}, current step name ${currentStepName}. Stay on current step.`);
    boundWorkflowRequestEnd();
  }
};

export const completeTask =
    ( application?: Application, currentStep?: string, isBackButton?: boolean, targetStep?: WORKFLOW_STEP_NAME, jobId?: string, schedule?: Schedule ) => {
      log("[WS] in completeTask, websocket is: ", {
        websocket: window.stepFunctionService?.websocket,
        readyState: window.stepFunctionService?.websocket?.readyState,
      });
      if (window.stepFunctionService?.websocket) {
        boundWorkflowRequestStart();
        const jobSelectedOn = application?.jobSelected?.jobSelectedOn || application?.jobScheduleSelected?.jobScheduleSelectedTime || "";
        const state = schedule?.state || "";
        const employmentType = schedule?.employmentType || "";
        const data: CompleteTaskRequest = {
          action: "completeTask",
          applicationId: window.stepFunctionService.applicationId || "",
          candidateId: window.stepFunctionService.candidateId || "",
          requisitionId: window.stepFunctionService.requisitionId || "", // requisitionId can't be null
          jobId: jobId || "",
          state,
          employmentType,
          eventSource: "HVH-CA-UI",
          jobSelectedOn,
          currentWorkflowStep: currentStep || "",
          isCsDomain: checkIfIsCSRequest(),
          workflowStepName: isBackButton ? targetStep || "" : ""
        };

        log(`[WS] complete ${currentStep} stepName request: `, { ...data });

        // wait for ready state to be open
        window.stepFunctionService.sendMessage(JSON.stringify(data));
      }
    };

export const onTimeOut = () => {
  if (window.hearBeatTime) {
    const endTime = moment();
    const startTime = moment(window.hearBeatTime);
    const duration = moment.duration(endTime.diff(startTime));

    if (duration.asMinutes() > MAX_MINUTES_FOR_HEARTBEAT) {
      log("Websocket timed out, moved to timed out page");

      boundWorkflowRequestEnd();
      routeToAppPageWithPath(PAGE_ROUTES.TIMEOUT);
    }
  } else {
    log("Websocket timed out, moved to timed out page");

    boundWorkflowRequestEnd();
    routeToAppPageWithPath(PAGE_ROUTES.TIMEOUT);
  }
};

export const onCompleteTaskHelper = ( application: Application, isBackButton?: boolean, targetStep?: WORKFLOW_STEP_NAME, currentStep?: WORKFLOW_STEP_NAME ) => {
  const state = store.getState();
  const jobId = application.jobScheduleSelected?.jobId;
  const scheduleId = application.jobScheduleSelected?.scheduleId;
  const { applicationId } = application;
  const { candidateId } = application;
  const currentStepName = currentStep || getCurrentStepNameFromHash();
  const { scheduleDetail } = state.schedule.results;

  if (targetStep) {
    window.hasTargetStep = true;
  }

  if (isBackButton) {
    log(`[WS] Completed task on back button execution, current step is ${currentStepName} for application:`, application);
  } else {
    log(`[WS] Completed task on ${currentStepName} for application:`, application);
  }

  if (!window?.stepFunctionService?.websocket && state.appConfig.results?.envConfig) {
    log("[WS] No websocket connection, add hasCompleteTaskOnWorkflowConnect and load workflow");
    window.hasCompleteTaskOnWorkflowConnect = () => {
      completeTask(application, currentStepName, isBackButton, targetStep, jobId, scheduleDetail);
    };
    boundWorkflowRequestStart();
    loadWorkflowDS(jobId || "", scheduleId || "", applicationId, candidateId, state.appConfig.results.envConfig);
  } else {
    completeTask(application, currentStepName, isBackButton, targetStep, jobId, scheduleDetail);
  }
};

export const actionWorkflowRequestInit = (): WorkflowRequestInitAction => {
  return { type: WORKFLOW_REQUEST.INIT };
};

export const actionWorkflowRequestStart = (): WorkflowRequestStartAction => {
  return { type: WORKFLOW_REQUEST.START };
};

export const actionWorkflowRequestEnd = (): WorkflowRequestEndAction => {
  return { type: WORKFLOW_REQUEST.END, loadingStatus: loadingStatusHelper() };
};

export const actionSetWorkflowErrorCode = (payload: WORKFLOW_ERROR_CODE): SetWorkflowErrorCodeAction => {
  return {
    type: WORKFLOW_REQUEST.SET_WORKFLOW_ERROR_CODE,
    payload
  };
};
