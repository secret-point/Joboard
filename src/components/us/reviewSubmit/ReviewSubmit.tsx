import React, { useEffect } from "react";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import { IconPencil } from "@amzn/stencil-react-components/icons";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { Text } from "@amzn/stencil-react-components/text";
import queryString from "query-string";
import { connect } from "react-redux";
import { useLocation } from "react-router-dom";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { boundUpdateApplicationDS } from "../../../actions/ApplicationActions/boundApplicationActions";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { boundResetBannerMessage } from "../../../actions/UiActions/boundUi";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import { ApplicationState } from "../../../reducers/application.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { JobState } from "../../../reducers/job.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { UpdateApplicationRequestDS } from "../../../utils/apiTypes";
import { CommonColors } from "../../../utils/colors";
import { FEATURE_FLAG, UPDATE_APPLICATION_API_TYPE, WORKFLOW_STEP_NAME } from "../../../utils/enums/common";
import { checkAndBoundGetApplication, createUpdateApplicationRequest, formatDate, getFeatureFlagValue, getLocale, reverseMappingTranslate } from "../../../utils/helper";
import { translate as t } from "../../../utils/translator";
import { Application } from "../../../utils/types/common";
import DebouncedButton from "../../common/DebouncedButton";
import ScheduleDetails from "../../common/jobOpportunity/ScheduleDetails";

interface MapStateToProps {
  job: JobState;
  application: ApplicationState;
  schedule: ScheduleState;
  candidate: CandidateState;
}

export const ReviewSubmit = (props: MapStateToProps) => {
  const { job, application, schedule, candidate } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId, scheduleId } = queryParams;

  const jobDetail = job.results;
  const { candidateData } = candidate.results;
  const { scheduleDetail } = schedule.results;

  const applicationData = application.results;
  const nheAppointment = applicationData?.nheAppointment;
  const location = applicationData?.nheAppointment?.location;

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    })
  }, [scheduleDetail, scheduleId]);

  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() })
  }, [jobDetail, jobId]);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    jobDetail && applicationData && candidateData && scheduleDetail && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail, pageName]);

  useEffect(() => {
    return () => {
      //reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    }
  },[pageName]);

  const handleBackToEdit = (stepName: WORKFLOW_STEP_NAME) => {
    boundResetBannerMessage();

    const isBackButton = true;
    if (applicationData) {
      onCompleteTaskHelper(applicationData, isBackButton, stepName);
    }
  };

  const handleSubmitApplication = () => {
    boundResetBannerMessage();

    if (applicationData) {
      const { REVIEW_SUBMIT } = UPDATE_APPLICATION_API_TYPE;
      const payload = {
        jobId,
        scheduleId,
        scheduleDetails: JSON.stringify(scheduleDetail)
      }

      if (getFeatureFlagValue(FEATURE_FLAG.BACKEND_HARD_RESERVE_SCHEDULE)) {
          onCompleteTaskHelper(applicationData);
      } else {
        const request: UpdateApplicationRequestDS = createUpdateApplicationRequest(applicationData, REVIEW_SUBMIT, payload);
        boundUpdateApplicationDS(request, (applicationData: Application) => {
          onCompleteTaskHelper(applicationData);
        });
      }
    }
  };

  return (
    <Col gridGap="S300" padding={{ top: "S300" }}>
      <Row gridGap="S300" padding="S500" style={{ background: `${CommonColors.Blue05}` }}>
        <Text fontSize="T300">
          {jobDetail?.jobTitle || ""}
        </Text>
      </Row>

      <Text fontSize="T400">
        {t("BB-ReviewSubmit-header-title-text", "Review and Submit")}
      </Text>

      <div>
        <Col gridGap="S300">
          <Text fontSize="T300">
            {t("BB-ReviewSubmit-job-details-section-title-text", "Job Details")}
          </Text>
          {scheduleDetail && <ScheduleDetails scheduleDetail={scheduleDetail} />}
        </Col>
      </div>

      <div>
        <Row alignItems="center" justifyContent="space-between">
          <Text fontSize="T300">
            {t("BB-ReviewSubmit-job-details-section-background-check-text", "Background check")}
          </Text>

          <Button icon={<IconPencil aria-hidden={true} />} variant={ButtonVariant.Tertiary} onClick={() => handleBackToEdit(WORKFLOW_STEP_NAME.BGC)}>
            {t("BB-ReviewSubmit-edit-button-text", "Edit")}
          </Button>
        </Row>

        <Col gridGap="S300">
          <Text fontSize="T100" fontWeight="bold">
            {t("BB-ReviewSubmit-job-details-section-authorized-text", "Authorized")}
          </Text>

          {applicationData?.nonFcraQuestions?.requestedCopyOfBackgroundCheck && <Text fontSize="T100">
            {t("BB-ReviewSubmit-job-details-section-send-free-copy-of-background-check-text", "Send free copy of background check")}
          </Text>}

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-job-details-section-national-id-type-text", "National ID Type:")} {reverseMappingTranslate(candidateData?.additionalBackgroundInfo.governmentIdType)}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-job-details-section-national-id-number-text", "National ID Number:")} {candidateData?.additionalBackgroundInfo.idNumber}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-job-details-section-date-of-birth-text", "Date of birth:")} {formatDate(candidateData?.additionalBackgroundInfo.dateOfBirth, { displayFormat: "Do MMM YYYY" })}
          </Text>
        </Col>
      </div>

      <div>
        <Row alignItems="center" justifyContent="space-between">
          <Text fontSize="T300">
            {t("BB-ReviewSubmit-pre-hire-appointment-section-title-text", "Pre-Hire Appointment")}
          </Text>

          <Button icon={<IconPencil aria-hidden={true} />} variant={ButtonVariant.Tertiary} onClick={() => handleBackToEdit(WORKFLOW_STEP_NAME.NHE)}>
            {t("BB-ReviewSubmit-edit-button-text", "Edit")}
          </Button>
        </Row>
        <Col gridGap="S300">
          <Text>
            {formatDate(nheAppointment?.dateWithoutFormat, {
              defaultDateFormat: "DD/MM/yyyy",
              displayFormat: "dddd, MMM Do YYYY"
            })}
          </Text>

          <Text fontSize="T100">
            {location && `${location.streetAddress} ${location.city} ${location.state} ${location.postalCode}`}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-pre-hire-appointment-section-hours-text", "Hours:")} {nheAppointment?.startTime} - {nheAppointment?.endTime}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-pre-hire-appointment-section-look-forward-text", "We look forward to seeing you at the time above.")}
          </Text>
        </Col>
      </div>

      <div>
        <Row alignItems="center" justifyContent="space-between">
          <Text fontSize="T300">
            {t("BB-ReviewSubmit-selfId-section-title-text", "Voluntary Self-identification")}
          </Text>

          <Button icon={<IconPencil aria-hidden={true} />} variant={ButtonVariant.Tertiary} onClick={() => handleBackToEdit(WORKFLOW_STEP_NAME.SELF_IDENTIFICATION)}>
            {t("BB-ReviewSubmit-edit-button-text", "Edit")}
          </Button>
        </Row>
        <Col gridGap="S300">
          <Text fontSize="T100">
            {t("BB-ReviewSubmit-selfId-section-equal-opportunity-text", "Equal Opportunity:")} {reverseMappingTranslate(candidateData?.selfIdentificationInfo.gender)}
            {candidateData?.selfIdentificationInfo.ethnicity ? `, ${reverseMappingTranslate(candidateData?.selfIdentificationInfo.ethnicity)}` : ""}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-selfId-section-veteral-status-text", "Veteral Status:")} {reverseMappingTranslate(candidateData?.selfIdentificationInfo.veteran)}
          </Text>

          <Text fontSize="T100">
            {t("BB-ReviewSubmit-selfId-section-disability-status-text", "Disability Status:")} {reverseMappingTranslate(candidateData?.selfIdentificationInfo.disability)}
          </Text>
        </Col>
      </div>

      <Col gridGap="S300" padding="S500" style={{ background: `${CommonColors.Blue05}` }}>
        <DebouncedButton variant={ButtonVariant.Primary} onClick={handleSubmitApplication}>
          {t("BB-ReviewSubmit-submit-application-button-text", "Submit application")}
        </DebouncedButton>
      </Col>
    </Col >
  )
}

const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(ReviewSubmit);
