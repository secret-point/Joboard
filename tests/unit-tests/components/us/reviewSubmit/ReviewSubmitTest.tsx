import React from "react";
import { shallow } from "enzyme";
import { useLocation } from 'react-router-dom';
import { ReviewSubmit } from "../../../../../src/components/us/reviewSubmit/ReviewSubmit";
import { TEST_APPLICATION_ID, TEST_APPLICATION_STATE, TEST_CANDIDATE_STATE, TEST_JOB_ID, TEST_JOB_STATE, TEST_SCHEDULE_ID, TEST_SCHEDULE_STATE } from "../../../../test-utils/test-data";
import { act } from "react-dom/test-utils";
import * as boundUi from "../../../../../src/actions/UiActions/boundUi";
import * as workflowActions from "../../../../../src/actions/WorkflowActions/workflowActions";
import { CountryCode } from "../../../../../src/utils/enums/common";
import * as utilHelpers from "../../../../../src/utils/helper";

const getCountryCodeSpy = jest.spyOn(utilHelpers, "getCountryCode");

describe("ReviewSubmit", () => {
  const mockLocation = {
    pathname: "/review-submit",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: '',
    state: null
  };

  beforeEach(() => {
    getCountryCodeSpy.mockReturnValue(CountryCode.US);
  })

  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue(mockLocation);

  it("should match snapshot", () => {
    const shallowWrapper = shallow(
      <ReviewSubmit
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });

  it("should call boundResetBannerMessage", () => {
    const shallowWrapper = shallow(
      <ReviewSubmit
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
      />);

    const spy = jest.spyOn(boundUi, 'boundResetBannerMessage');

    const submitButton = shallowWrapper.find('DebouncedButton').first();

    act(() => {
      submitButton.simulate('click');
    })

    expect(spy).toHaveBeenCalled();
  });

  it("should call boundResetBannerMessage", () => {
    const shallowWrapper = shallow(
      <ReviewSubmit
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
      />);

    const spy = jest.spyOn(workflowActions, 'onCompleteTaskHelper');

    const editButtons = shallowWrapper.find('Button');

    act(() => {
      editButtons.forEach(button => button.simulate('click'));
    })

    expect(spy).toHaveBeenCalledTimes(editButtons.length);
  });
});

