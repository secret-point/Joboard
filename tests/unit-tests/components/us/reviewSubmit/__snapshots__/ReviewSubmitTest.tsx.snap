// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewSubmit should match snapshot 1`] = `
<Col
  gridGap="S300"
  padding={
    Object {
      "top": "S300",
    }
  }
>
  <Row
    gridGap="S300"
    padding="S500"
    style={
      Object {
        "background": "#EBF5FD",
      }
    }
  >
    <Text
      font="primary"
      fontSize="T300"
      hyphens="none"
    />
  </Row>
  <Text
    font="primary"
    fontSize="T400"
    hyphens="none"
  >
    Review and Submit
  </Text>
  <div>
    <Col
      gridGap="S300"
    >
      <Text
        font="primary"
        fontSize="T300"
        hyphens="none"
      >
        Job Details
      </Text>
      <ScheduleDetails
        scheduleDetail={
          Object {
            "scheduleId": "test-schedule-id",
          }
        }
      />
    </Col>
  </div>
  <div>
    <Row
      alignItems="center"
      justifyContent="space-between"
    >
      <Text
        font="primary"
        fontSize="T300"
        hyphens="none"
      >
        Background check
      </Text>
      <Button
        colorScheme="default"
        icon={
          <IconPencil
            display="block"
          />
        }
        iconPosition="leading"
        onClick={[Function]}
        shouldUseResponsiveContent={true}
        size="default"
        type="button"
        variant="tertiary"
      >
        Edit
      </Button>
    </Row>
    <Col
      gridGap="S300"
    >
      <Text
        font="primary"
        fontSize="T100"
        fontWeight="bold"
        hyphens="none"
      >
        Authorized
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        National ID Type:
         
        United States - Social Security Number
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        National ID Number:
         
        *****9599
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        Date of birth:
         
        2nd Aug 1988
      </Text>
    </Col>
  </div>
  <div>
    <Row
      alignItems="center"
      justifyContent="space-between"
    >
      <Text
        font="primary"
        fontSize="T300"
        hyphens="none"
      >
        Pre-Hire Appointment
      </Text>
      <Button
        colorScheme="default"
        icon={
          <IconPencil
            display="block"
          />
        }
        iconPosition="leading"
        onClick={[Function]}
        shouldUseResponsiveContent={true}
        size="default"
        type="button"
        variant="tertiary"
      >
        Edit
      </Button>
    </Row>
    <Col
      gridGap="S300"
    >
      <Text
        font="primary"
        fontSize="T300"
        hyphens="none"
      >
        Invalid date
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      />
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        Hours:
         
         - 
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        We look forward to seeing you at the time above.
      </Text>
    </Col>
  </div>
  <div>
    <Row
      alignItems="center"
      justifyContent="space-between"
    >
      <Text
        font="primary"
        fontSize="T300"
        hyphens="none"
      >
        Voluntary Self-identification
      </Text>
      <Button
        colorScheme="default"
        icon={
          <IconPencil
            display="block"
          />
        }
        iconPosition="leading"
        onClick={[Function]}
        shouldUseResponsiveContent={true}
        size="default"
        type="button"
        variant="tertiary"
      >
        Edit
      </Button>
    </Row>
    <Col
      gridGap="S300"
    >
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        Equal Opportunity:
         
        Male
        , Asian (not Hispanic or Latino)
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        Veteral Status:
         
        No
      </Text>
      <Text
        font="primary"
        fontSize="T100"
        hyphens="none"
      >
        Disability Status:
         
      </Text>
    </Col>
  </div>
  <Col
    gridGap="S300"
    padding="S500"
    style={
      Object {
        "background": "#EBF5FD",
      }
    }
  >
    <Button
      colorScheme="default"
      iconPosition="leading"
      onClick={[Function]}
      shouldUseResponsiveContent={true}
      size="default"
      type="button"
      variant="primary"
    >
      Submit application
    </Button>
  </Col>
</Col>
`;
